cosmovisor/process.go

```diff
@@ -143,3 +151,65 @@ func doBackup(cfg *Config) error {
 
 	return nil
 }
+
+// doPreUpgrade runs the pre-upgrade command defined by the application
+func doPreUpgrade(cfg *Config) error {
+	bin, err := cfg.CurrentBin()
+	preUpgradeCmd := exec.Command(bin, "pre-upgrade")
+
+	_, err = preUpgradeCmd.Output()
+
+	if err != nil {
+		if err.(*exec.ExitError).ProcessState.ExitCode() == 1 {
+			fmt.Println("pre-upgrade command does not exist. continuing the upgrade.")
+			return nil
+		}
+		if err.(*exec.ExitError).ProcessState.ExitCode() == 30 {
+			return fmt.Errorf("pre-upgrade command failed : %w", err)
+		}
+		if err.(*exec.ExitError).ProcessState.ExitCode() == 31 {
+			fmt.Println("pre-upgrade command failed. retrying.")
+			return doPreUpgrade(cfg)
+		}
+	}
+	fmt.Println("pre-upgrade successful. continuing the upgrade.")
+	return nil
+}
+
+// skipUpgrade checks if pre-upgrade script must be run. If the height in the upgrade plan matches any of the heights provided in --safe-skip-upgrade, the script is not run
+func SkipUpgrade(args []string, upgradeInfo UpgradeInfo) bool {
```

Comment:
This naming is bit ambiguous. How about `isSkipUpgradeHeight` or something better. Also why are we exporting this function?